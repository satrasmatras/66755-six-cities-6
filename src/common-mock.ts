import Offer from "./models/offer";
import Comment from "./models/comment";
import City from "./models/city";
import SortType from "./models/sort-type";
import Routes from "./routes";
import {AuthorizationStatus, LoginPayload} from "./store/user/slice";
import AuthInfo from "./models/auth-info";
import {adaptDataToOffer} from "./adapters/offers";
import {adaptDataToComment} from "./adapters/comments";
import {CommentPost} from "./models/comment-post";
import {adaptDataToAuthInfo} from "./adapters/auth-info";
import {RootState} from "./store";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";
import {redirectMiddleware} from "./store/redirect";

export const MOCK_OFFER_FROM_API = {
  "bedrooms": 3,
  "city": {
    "location": {
      "latitude": 52.370216,
      "longitude": 4.895168,
      "zoom": 10
    },
    "name": `Amsterdam`
  },
  "description": `A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam.`,
  "goods": [`Heating`, `Kitchen`, `Cable TV`, `Washing machine`, `Coffee machine`, `Dishwasher`],
  "host": {
    "avatar_url": `img/1.png`,
    "id": 3,
    "is_pro": true,
    "name": `Angelina`
  },
  "id": 1,
  "images": [`img/1.png`, `img/2.png`],
  "is_favorite": false,
  "is_premium": false,
  "location": {
    "latitude": 52.35514938496378,
    "longitude": 4.673877537499948,
    "zoom": 8
  },
  "max_adults": 4,
  "preview_image": `img/1.png`,
  "price": 120,
  "rating": 4.8,
  "title": `Beautiful & luxurious studio at great location`,
  "type": `apartment`
};
export const MOCK_ADAPTED_OFFER: Offer = adaptDataToOffer(MOCK_OFFER_FROM_API);

export const MOCK_OFFERS_FROM_API = [
  MOCK_OFFER_FROM_API,
  MOCK_OFFER_FROM_API,
  MOCK_OFFER_FROM_API,
  MOCK_OFFER_FROM_API
].map((offer, i) => ({
  ...offer,
  id: i + 1
}));

export const MOCK_ADAPTED_OFFERS: Offer[] = MOCK_OFFERS_FROM_API.map(adaptDataToOffer);

export const MOCK_COMMENT_FROM_API = {
  "comment": `A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam.`,
  "date": `2019-05-08T14:13:56.569Z`,
  "id": 1,
  "rating": 4,
  "user": {
    "avatar_url": `img/1.png`,
    "id": 4,
    "isPro": false,
    "name": `Max`
  }
};

export const MOCK_COMMENTS_FROM_API = [
  MOCK_COMMENT_FROM_API,
  MOCK_COMMENT_FROM_API,
  MOCK_COMMENT_FROM_API,
  MOCK_COMMENT_FROM_API
].map((comment, i) => ({
  ...comment,
  id: i + 1
}));

export const MOCK_ADAPTED_COMMENT: Comment = adaptDataToComment(MOCK_COMMENT_FROM_API);

export const MOCK_ADAPTED_COMMENTS: Comment[] = MOCK_COMMENTS_FROM_API.map(adaptDataToComment);

export const MOCK_CITY: City = {
  "location": {
    "latitude": 52.370216,
    "longitude": 4.895168,
    "zoom": 10
  },
  "name": `Amsterdam`
};

export const MOCK_SORT_TYPE = SortType.TOP_RATED;

export const MOCK_ROUTE = Routes.FAVORITES;

export const MOCK_AUTH_INFO_FROM_API = {
  "avatar_url": `img/1.png`,
  "email": `Oliver.conner@gmail.com`,
  "id": 1,
  "is_pro": false,
  "name": `Oliver.conner`
};

export const MOCK_ADAPTED_AUTH_INFO: AuthInfo = adaptDataToAuthInfo(MOCK_AUTH_INFO_FROM_API);

export const MOCK_AUTHORIZATION_STATUS = AuthorizationStatus.AUTH;

export const MOCK_EMPTY_ACTION = {
  type: `test`,
  payload: 0
};

export const MOCK_LOGIN_PAYLOAD: LoginPayload = {
  email: MOCK_ADAPTED_AUTH_INFO.email,
  password: `12345`
};

export const MOCK_COMMENT_POST: CommentPost = {
  comment: `Lalalalalalallaallaa`,
  rating: 5,
};

export const MOCK_INITIAL_STATE: RootState = {
  offers: {
    offers: MOCK_ADAPTED_OFFERS,
    isLoading: false,
    sortType: MOCK_SORT_TYPE
  },
  city: {
    city: MOCK_CITY
  },
  map: {
    hoveredOffer: MOCK_ADAPTED_OFFER
  },
  user: {
    authorizationStatus: MOCK_AUTHORIZATION_STATUS,
    authInfo: MOCK_ADAPTED_AUTH_INFO
  },
  offer: {
    offer: MOCK_ADAPTED_OFFER,
    offerIsLoading: false,
    comments: MOCK_ADAPTED_COMMENTS,
    commentsAreLoading: false
  },
  favorites: {
    favorites: MOCK_ADAPTED_OFFERS,
    isLoading: false
  }
};

export const MOCK_STORE = configureStore([thunk, redirectMiddleware])(MOCK_INITIAL_STATE);
