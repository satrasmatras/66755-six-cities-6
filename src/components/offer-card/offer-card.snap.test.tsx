import React from "react";
import {render} from "@testing-library/react";
import OfferCard from "./offer-card";
import OfferCardTypes from "../../models/offer-card-types";
import {MOCK_ADAPTED_OFFER, MOCK_INITIAL_STATE, mockConfigureStore} from "../../common-mock";
import {Router} from "react-router-dom";
import browserHistory from "../../browser-history";
import * as redux from "react-redux";
import {Provider} from "react-redux";
import {createMemoryHistory} from "history";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";

describe(`offer card should render correctly`, () => {
  const mockFactory = configureStore([thunk]);

  const useSelectorMock = jest.spyOn(redux, `useSelector`);
  const useDispatchMock = jest.spyOn(redux, `useDispatch`);

  beforeEach(() => {
    useSelectorMock.mockClear();
    useDispatchMock.mockClear();

    useDispatchMock.mockReturnValue(jest.fn());
  });

  it(`Should main card render correctly`, () => {
    const {container} = render(
        <Provider store={mockFactory(MOCK_INITIAL_STATE)}>
          <Router history={browserHistory}>
            <OfferCard
              cardType={OfferCardTypes.CITY}
              key={1}
              offer={MOCK_ADAPTED_OFFER}
              handleHover={jest.fn()}
              handleBookmark={jest.fn()}
            />
          </Router>
        </Provider>
    );
    expect(container).toMatchSnapshot();
  });

  it(`Should favorite render correctly`, () => {
    const {container} = render(
        <Provider store={mockFactory(MOCK_INITIAL_STATE)}>
          <Router history={browserHistory}>
            <OfferCard
              cardType={OfferCardTypes.FAVORITE}
              key={1}
              offer={MOCK_ADAPTED_OFFER}
              handleHover={jest.fn()}
              handleBookmark={jest.fn()}
            />
          </Router>
        </Provider>
    );
    expect(container).toMatchSnapshot();
  });

  it(`Should nearby render correctly`, () => {
    const {container} = render(
        <Provider store={mockFactory(MOCK_INITIAL_STATE)}>
          <Router history={browserHistory}>
            <OfferCard
              cardType={OfferCardTypes.NEAR}
              key={1}
              offer={MOCK_ADAPTED_OFFER}
              handleHover={jest.fn()}
              handleBookmark={jest.fn()}
            />
          </Router>
        </Provider>
    );
    expect(container).toMatchSnapshot();
  });
});
