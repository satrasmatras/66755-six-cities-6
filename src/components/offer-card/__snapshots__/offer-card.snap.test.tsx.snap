// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should favorite render correctly 1`] = `
<div>
  <article
    class="favorites__card place-card"
  >
    <div
      class="favorites__image-wrapper place-card__image-wrapper"
    >
      <a
        href="/offer/1"
      >
        <img
          alt="Place image"
          class="place-card__image"
          height="110"
          src="img/1.png"
          width="150"
        />
      </a>
    </div>
    <div
      class="favorites__card-info place-card__info"
    >
      <div
        class="place-card__price-wrapper"
      >
        <div
          class="place-card__price"
        >
          <b
            class="place-card__price-value"
          >
            €
            120
          </b>
          <span
            class="place-card__price-text"
          >
            / night
          </span>
        </div>
        <button
          class="place-card__bookmark-button  button"
          type="button"
        >
          <svg
            class="place-card__bookmark-icon"
            height="19"
            width="18"
          >
            <use
              xlink:href="#icon-bookmark"
            />
          </svg>
          <span
            class="visually-hidden"
          >
            To bookmarks
          </span>
        </button>
      </div>
      <div
        class="place-card__rating rating"
      >
        <div
          class="place-card__stars rating__stars"
        >
          <span
            style="width: 100%;"
          />
          <span
            class="visually-hidden"
          >
            Rating
          </span>
        </div>
      </div>
      <h2
        class="place-card__name"
      >
        <a
          href="/offer/1"
        >
          Beautiful & luxurious studio at great location
        </a>
      </h2>
      <p
        class="place-card__type"
      >
        Apartment
      </p>
    </div>
  </article>
</div>
`;

exports[`Should main card render correctly 1`] = `
<div>
  <article
    class="cities__place-card place-card"
  >
    <div
      class="cities__image-wrapper place-card__image-wrapper"
    >
      <a
        href="/offer/1"
      >
        <img
          alt="Place image"
          class="place-card__image"
          height="200"
          src="img/1.png"
          width="260"
        />
      </a>
    </div>
    <div
      class=" place-card__info"
    >
      <div
        class="place-card__price-wrapper"
      >
        <div
          class="place-card__price"
        >
          <b
            class="place-card__price-value"
          >
            €
            120
          </b>
          <span
            class="place-card__price-text"
          >
            / night
          </span>
        </div>
        <button
          class="place-card__bookmark-button  button"
          type="button"
        >
          <svg
            class="place-card__bookmark-icon"
            height="19"
            width="18"
          >
            <use
              xlink:href="#icon-bookmark"
            />
          </svg>
          <span
            class="visually-hidden"
          >
            To bookmarks
          </span>
        </button>
      </div>
      <div
        class="place-card__rating rating"
      >
        <div
          class="place-card__stars rating__stars"
        >
          <span
            style="width: 100%;"
          />
          <span
            class="visually-hidden"
          >
            Rating
          </span>
        </div>
      </div>
      <h2
        class="place-card__name"
      >
        <a
          href="/offer/1"
        >
          Beautiful & luxurious studio at great location
        </a>
      </h2>
      <p
        class="place-card__type"
      >
        Apartment
      </p>
    </div>
  </article>
</div>
`;

exports[`Should nearby render correctly 1`] = `
<div>
  <article
    class="near-places__card place-card"
  >
    <div
      class="near-places__image-wrapper place-card__image-wrapper"
    >
      <a
        href="/offer/1"
      >
        <img
          alt="Place image"
          class="place-card__image"
          src="img/1.png"
        />
      </a>
    </div>
    <div
      class=" place-card__info"
    >
      <div
        class="place-card__price-wrapper"
      >
        <div
          class="place-card__price"
        >
          <b
            class="place-card__price-value"
          >
            €
            120
          </b>
          <span
            class="place-card__price-text"
          >
            / night
          </span>
        </div>
        <button
          class="place-card__bookmark-button  button"
          type="button"
        >
          <svg
            class="place-card__bookmark-icon"
            height="19"
            width="18"
          >
            <use
              xlink:href="#icon-bookmark"
            />
          </svg>
          <span
            class="visually-hidden"
          >
            To bookmarks
          </span>
        </button>
      </div>
      <div
        class="place-card__rating rating"
      >
        <div
          class="place-card__stars rating__stars"
        >
          <span
            style="width: 100%;"
          />
          <span
            class="visually-hidden"
          >
            Rating
          </span>
        </div>
      </div>
      <h2
        class="place-card__name"
      >
        <a
          href="/offer/1"
        >
          Beautiful & luxurious studio at great location
        </a>
      </h2>
      <p
        class="place-card__type"
      >
        Apartment
      </p>
    </div>
  </article>
</div>
`;
